packages/rpc-provider/src/substrate-connect/createScClient.ts

```diff
@@ -0,0 +1,348 @@
+        if (this.#isChainReady === isReady) return;
+
+        this.#isChainReady = isReady;
+
+        if (!isReady) {
+          // If we've reached this point, that means that the chain used to be "ready"
+          // and now we are about to emit `disconnected`.
+          //
+          // This will cause the PolkadotJs API think that the connection is
+          // actually dead. In reality the smoldot chain is not dead, of course.
+          // However, we have to cleanup all the existing callbacks because when
+          // the smoldot chain stops syncing, then we will emit `connected` and
+          // the PolkadotJs API will try to re-create the previous
+          // subscriptions and requests. Although, now is not a good moment
+          // to be sending unsubscription messages to the smoldot chain, we
+          // should wait until is no longer syncing to send the unsubscription
+          // messages from the stale subscriptions of the previous connection.
+          //
+          // That's why -before we perform the cleanup of `this.#subscriptions`-
+          // we keep the necessary information that we will need later on to
+          // kill the stale subscriptions.
+          [...this.#subscriptions.values()].forEach((s) => {
+            staleSubscriptions.push(s[1]);
+          });
+          cleanup();
+        } else {
+          killStaleSubscriptions();
+        }
+
+        this.#eventemitter.emit(isReady ? 'connected' : 'disconnected');
+      });
+
+      return {
+        ...chain,
+        remove: () => {
+          hc.stop();
+          chain.remove();
+          cleanup();
+        },
+        sendJsonRpc: hc.sendJsonRpc.bind(hc)
+      };
+    });
+
+    try {
+      await this.#chain;
+    } catch (e) {
+      this.#chain = null;
+      this.#eventemitter.emit('error', e);
+      throw e;
+    }
+  }
+
+  async disconnect (): Promise<void> {
+    if (!this.#chain) return;
+
+    const chain = await this.#chain;
+
+    this.#chain = null;
+    this.#isChainReady = false;
+
+    try {
+      chain.remove();
+    } catch (_) {}
+
+    this.#eventemitter.emit('disconnected');
+  }
+
+  public on (
+    type: ProviderInterfaceEmitted,
+    sub: ProviderInterfaceEmitCb
+  ): () => void {
+    // It's possible. Although, quite unlikely, that by the time that polkadot
+    // subscribes to the `connected` event, the Provider is already connected.
+    // In that case, we must emit to let the consumer know that we are connected.
+    if (type === 'connected' && this.isConnected) {
+      sub();
+    }
+
+    this.#eventemitter.on(type, sub);
+
+    return (): void => {
+      this.#eventemitter.removeListener(type, sub);
+    };
+  }
+
+  public async send<T = any> (method: string, params: unknown[]): Promise<T> {
+    if (!this.isConnected) throw new Error('Provider is not connected');
+
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const chain = await this.#chain!;
```

Comment:
Not really, this assertion is bogus, because the previous line (`if(!this.isConnected) throw...`, or `assert(this.isConnected, 'Provider is not connected')`) already ensures that the `chain` is present. But it's up to you...